src/main/java/ch/njol/skript/expressions/ExprIndicesOf.java

```diff
@@ -0,0 +1,243 @@
+	"",
+	"set {_list::*} to 1, 2, 3, 1, 2, 3",
+	"set {_indices::*} to all indices of the value 1 in {_list::*}",
+	"# {_indices::*} is now \"1\" and \"4\"",
+	"",
+	"set {_indices::*} to all indices of the value 2 in {_list::*}",
+	"# {_indices::*} is now \"2\" and \"5\"",
+	"",
+	"set {_positions::*} to all positions of the value 3 in {_list::*}",
+	"# {_positions::*} is now 3 and 6",
+	"",
+	"set {_otherlist::bar} to 100",
+	"set {_otherlist::hello} to \"hi\"",
+	"set {_otherlist::burb} to 100",
+	"set {_otherlist::tud} to \"hi\"",
+	"set {_otherlist::foo} to 100",
+	"",
+	"set {_indices::*} to the first index of the value 100 in {_otherlist::*}",
+	"# {_indices::*} is now \"bar\"",
+	"set {_indices::*} to the last index of the value 100 in {_otherlist::*}",
+	"# {_indices::*} is now \"foo\"",
+	"",
+	"set {_positions::*} to all positions of the value 100 in {_otherlist::*}",
+	"# {_positions::*} is now 1, 3 and 5",
+	"set {_positions::*} to all positions of the value \"hi\" in {_otherlist::*}",
+	"# {_positions::*} is now 2 and 4"
+})
+@Since("2.1, INSERT VERSION (indices, positions of list)")
+public class ExprIndicesOf extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprIndicesOf.class, Object.class, ExpressionType.COMBINED,
+			"[the] [first|1:last|2:all] (position[s]|indices|index[es]) of [[the] value] %string% in %string%",
+			"[the] [first|1:last|2:all] position[s] of [[the] value] %object% in %~objects%",
+			"[the] [first|1:last|2:all] (indices|index[es]) of [[the] value] %object% in %~objects%"
+		);
+	}
+	
+	private IndexType type;
+	private boolean position;
+	private boolean string;
+	private Expression<?> value, objects;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (exprs[1].isSingle() && (matchedPattern > 0)) {
+			Skript.error("'" + exprs[1] + "' can only ever have one value at most, thus the 'indices of x in list' expression has no effect.");
+			return false;
+		}
+
+		if (!(exprs[1] instanceof Variable<?>) && matchedPattern == 2) {
+			Skript.error("'" + exprs[1] + "' is not a list variable. You can only get the indices of a list variable.");
+			return false;
+		}
+
+		type = IndexType.values()[parseResult.mark];
+		position = matchedPattern <= 1;
+		string = matchedPattern == 0;
+		value = LiteralUtils.defendExpression(exprs[0]);
+		objects = exprs[1];
+
+		return LiteralUtils.canInitSafely(value);
+	}
+	
+	@Override
+	protected Object @Nullable [] get(Event event) {
+		Object value = this.value.getSingle(event);
+		if (value == null) {
+			return (Object[]) Array.newInstance(getReturnType(), 0);
+		}
+
+		if (this.position) {
+			if (string) {
+				String haystack = (String) objects.getSingle(event);
+				if (haystack == null) {
+					return new Long[0];
+				}
+
+				return getStringPositions(haystack, (String) value);
+			}
+
+			Object[] list = objects.getArray(event);
```

Comment:
You should use iterators (`Expression#iterator`) to avoid unnecessary computations